### PR TITLE
[FG:InPlacePodVerticalScaling] Enable resizing containers without limits

### LINKED ISSUE #126388: [FG:InPlacePodVerticalScaling] Resizing pod gets stuck if limit is not configured
### What happened?

Resizing CPU requests gets stuck in `InProgress` if the CPU limit is not configured:
```
$ kubectl patch pod resize-pod --patch '{"spec":{"containers":[{"name":"resize-container", "resources":{"requests":{"cpu":"300m"}}}]}}'
pod/resize-pod patched
$ sleep 300
$ kubectl get pod resize-pod -o jsonpath='spec: {.spec.containers[0].resources}{"\nallocatedResources: "}{.status.containerStatuses[0].allocatedResources}{"\nstatus: "}{.status.containerStatuses[0].resources}{"\nresize: "}{.status.resize}{"\n"}'
spec: {"requests":{"cpu":"300m","memory":"200Mi"}}
allocatedResources: {"cpu":"300m","memory":"200Mi"}
status: {"requests":{"cpu":"200m","memory":"200Mi"}}
resize: InProgress
```

The internal behavior in kubelet varies depending on other resource limits configuration.
- If the limit for another resource is also not configured, no action is computed at `computePodResizeAction()`:
  ```
  Jul 26 13:43:09 kind-control-plane kubelet[708]: I0726 13:43:09.494698     708 kuberuntime_manager.go:1051] "computePodActions got for pod" podActions="KillPod: false, CreateSandbox: false, UpdatePodResources: false, Attempt: 0, InitContainersToStart: [], ContainersToStart: [], EphemeralContainersToStart: [],ContainersToUpdate: map[], ContainersToKill: map[]" pod="default/resize-pod"
  ```
  This looks caused because `container.Resources.Limits == nil`:
  https://github.com/kubernetes/kubernetes/blob/3a8a60eba29940e26ac8db52329a91ba87305114/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L555-L557
- If the limit for another resource is configured, `doPodResizeAction()` gets failed:
  ```
  Jul 26 14:04:23 kind-control-plane kubelet[708]: E0726 14:04:23.063108     708 kuberuntime_manager.go:750] "podResources.CPUQuota or podResources.CPUShares is nil" pod="resize-pod"
  ```
  This message is logged here:
  https://github.com/kubernetes/kubernetes/blob/3a8a60eba29940e26ac8db52329a91ba87305114/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L754-L757
  This is caused because CPUQuota is not set if the CPU limit is not configured:
  https://github.com/kubernetes/kubernetes/blob/3a8a60eba29940e26ac8db52329a91ba87305114/pkg/kubelet/cm/helpers_linux.go#L179-L182

There is a similar problem in resizing memory limits. See the comment below: https://github.com/kubernetes/kubernetes/issues/126388#issuecomment-2276385213

### What did you expect to happen?

This resizing should not get stuck in `InProgress` though I’m not sure whether this should be actuated or rejected yet.

### How can we reproduce it (as minimally and precisely as possible)?

1. Enable `InPlacePodVerticalScaling`.
2. Create a pod where CPU request is configured and CPU limit is not configured:
    <Details>

   pod with no resource limits:
   ```
   apiVersion: v1
   kind: Pod
   metadata:
     creationTimestamp: null
     labels:
       run: resize-pod
     name: resize-pod
   spec:
     containers:
     - image: busybox
       name: resize-container
       command:
         - sh
         - -c
         - trap "exit 0" SIGTERM; while true; do sleep 1; done
       resources:
         requests:
           cpu: 200m
           memory: 200Mi
       resizePolicy:
       - resourceName: cpu
         restartPolicy: NotRequired
       - resourceName: memory
         restartPolicy: NotRequired
     restartPolicy: Always
   ```

   pod with memory limit:
   ```
   apiVersion: v1
   kind: Pod
   metadata:
     creationTimestamp: null
     labels:
       run: resize-pod
     name: resize-pod
   spec:
     containers:
     - image: busybox
       name: resize-container
       command:
         - sh
         - -c
         - trap "exit 0" SIGTERM; while true; do sleep 1; done
       resources:
         requests:
           cpu: 200m
           memory: 200Mi
         limits:
           memory: 200Mi
       resizePolicy:
       - resourceName: cpu
         restartPolicy: NotRequired
       - resourceName: memory
         restartPolicy: NotRequired
     restartPolicy: Always
   ```

   </Details>
3. Patch the pod to update its CPU request:
   ```
   $ kubectl patch pod resize-pod --patch '{"spec":{"containers":[{"name":"resize-container", "resources":{"requests":{"cpu":"300m"}}}]}}'
   ```
4. Wait for more than two minutes in order to confirm this issue is not the [known issue](https://kubernetes.io/blog/2023/05/12/in-place-pod-resize-alpha/#known-issues).
5. See the pod resize status:
   ```
   $ kubectl get pod resize-pod -o jsonpath='spec: {.spec.containers[0].resources}{"\nallocatedResources: "}{.status.containerStatuses[0].allocatedResources}{"\nstatus: "}{.status.containerStatuses[0].resources}{"\nresize: "}{.status.resize}{"\n"}'
   spec: {"requests":{"cpu":"300m","memory":"200Mi"}}
   allocatedResources: {"cpu":"300m","memory":"200Mi"}
   status: {"requests":{"cpu":"200m","memory":"200Mi"}}
   resize: InProgress
   ```

### Anything else we need to know?

_No response_

### Kubernetes version

<details>

```console
$ kubectl version
Client Version: v1.30.3
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.30.0
```

</details>

### Cloud provider

N/A

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by tallclair**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

Several bugs in the resize code prevented containers without limits from resizing properly. This PR ensures that the cases where limits are missing are handled correctly, and adds test coverage for all these cases.

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/kubernetes/issues/126388

#### Special notes for your reviewer:

Extension request: https://groups.google.com/g/kubernetes-sig-release/c/DvRlV9mBZ-Q/m/AGmOk6avBwAJ

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

/sig node
/milestone v1.32
/priority important-soon

### CODE REVIEW

**Comment by tallclair** [comment]
/cc @hshiina 

Some of these changes overlap with https://github.com/kubernetes/kubernetes/pull/127143. I also used your `doPodResizeAction` test as the starting point for the unit test in this PR.

**Comment by hshiina** [comment]
It seems necessary to handle a case where CPU limit in the runtime status (`status.Resources.CPULimit`) is `nil` at a watch condition:
https://github.com/kubernetes/kubernetes/blob/1b08de5f7f964323f1e65a7db5e861429087efe6/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L827-L832

**Comment by tallclair** [comment]
Thanks. Fixed.

File: pkg/kubelet/kuberuntime/kuberuntime_manager.go [hunk 1]
```diff
@@ -584,8 +584,10 @@ func (m *kubeGenericRuntimeManager) computePodResizeAction(pod *v1.Pod, containe
 		cpuRequest:    container.Resources.Requests.Cpu().MilliValue(),
 	}
 
-	// Default current values to the desired values so that a resize isn't triggered for missing values.
-	currentResources := desiredResources
+	currentResources := containerResources{
+		// memoryRequest isn't set by the runtime, so default it to the desired.
+		memoryRequest: desiredResources.memoryRequest,
```
Comment (vinaykul) [inline pkg/kubelet/kuberuntime/kuberuntime_manager.go:589]: Shouldn't this come from checkpoint store?

Comment (vinaykul) [reply pkg/kubelet/kuberuntime/kuberuntime_manager.go:589]: NVM. This works after reading the whole function and refreshing my memory.

File: pkg/kubelet/kuberuntime/kuberuntime_manager_test.go [hunk 2]
```diff
@@ -2826,3 +2830,196 @@ func TestGetImageVolumes(t *testing.T) {
...
+			},
+			desiredResources: containerResources{
+				cpuRequest: 100, cpuLimit: 100,
+				memoryRequest: 100, memoryLimit: 100,
+			},
+			updatedResources:       []v1.ResourceName{v1.ResourceCPU, v1.ResourceMemory},
+			expectPodCgroupUpdates: 0,
+		},
+		{
+			testName: "Add limits and pod limits",
```
Comment (vinaykul) [inline pkg/kubelet/kuberuntime/kuberuntime_manager_test.go:2935]: Did you mean "Add cpu and memory limits"?

**Review by vinaykul — APPROVED** [review]
One minor comment.
LGTM

File: pkg/kubelet/kuberuntime/kuberuntime_manager_test.go [hunk 3]
```diff
@@ -2826,3 +2830,196 @@ func TestGetImageVolumes(t *testing.T) {
...
+			actions := podActions{
+				ContainersToUpdate: containersToUpdate,
+			}
+			m.doPodResizeAction(pod, actions, syncResult)
+
+			if tc.expectedError != "" {
+				assert.ErrorContains(t, syncResult.Error(), tc.expectedError)
+			} else {
+				assert.NoError(t, syncResult.Error())
+			}
```
Comment (esotsal) [inline pkg/kubelet/kuberuntime/kuberuntime_manager_test.go:3020]:
```suggestion
			if tc.expectedError != "" {
				require.Error(t, syncResult.Error())       
				require.EqualError(t, syncResult.Error(), tc.expectedError)
			} else {
				require.NoError(t, syncResult.Error())
			}
```

File: pkg/kubelet/kuberuntime/kuberuntime_manager_test.go (hunk 2 above)
Comment (dchen1107) [reply pkg/kubelet/kuberuntime/kuberuntime_manager_test.go:2935]: I believe the test case here is indeed update pod cgroup's limit, different from the previous test case.

File: pkg/kubelet/kuberuntime/kuberuntime_manager_test.go (hunk 3 above)
Comment (dchen1107) [reply pkg/kubelet/kuberuntime/kuberuntime_manager_test.go:3020]: Can you address this one? Otherwise we are ready to go.

Comment (tallclair) [reply pkg/kubelet/kuberuntime/kuberuntime_manager_test.go]: Done.

**Pull request merged**